builtin/providers/azurerm/resource_arm_virtual_network.go

```diff
@@ -0,0 +1,256 @@
+	log.Printf("[DEBUG] Waiting for Virtual Network (%s) to become available", d.Id())
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"Accepted", "Updating"},
+		Target:  "Succeeded",
+		Refresh: virtualNetworkStateRefreshFunc(client, resGroup, name),
+		Timeout: 10 * time.Minute,
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Error waiting for Virtual Network (%s) to become available: %s", d.Id(), err)
+	}
+
+	return resourceArmVirtualNetworkRead(d, meta)
+}
+
+// resourceArmVirtualNetworkRead goes ahead and reads the state of the corresponding ARM virtual network.
+func resourceArmVirtualNetworkRead(d *schema.ResourceData, meta interface{}) error {
+	vnetClient := meta.(*ArmClient).vnetClient
+
+	name := d.Id()
+	resGroup := d.Get("resource_group_name").(string)
+
+	log.Printf("[INFO] Sending virtual network read request to ARM.")
+
+	resp, err := vnetClient.Get(resGroup, name)
+	if resp.StatusCode == http.StatusNotFound {
+		// it means the virtual network has been deleted in the meantime;
+		// so we must go ahead and remove it here:
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Error making Read request on Azure virtual network %s: %s", name, err)
+	}
+	vnet := *resp.Properties
+
+	// update appropriate values
+	d.Set("address_space", vnet.AddressSpace.AddressPrefixes)
+
+	subnets := &schema.Set{
+		F: resourceAzureSubnetHash,
+	}
+
+	for _, subnet := range *vnet.Subnets {
+		s := map[string]interface{}{}
+
+		s["name"] = *subnet.Name
+		s["address_prefix"] = *subnet.Properties.AddressPrefix
+		if subnet.Properties.NetworkSecurityGroup != nil {
+			s["security_group"] = *subnet.Properties.NetworkSecurityGroup.ID
+		}
+
+		subnets.Add(s)
+	}
+	d.Set("subnet", subnets)
+
+	dnses := []string{}
+	for _, dns := range *vnet.DhcpOptions.DNSServers {
+		dnses = append(dnses, dns)
+	}
+	d.Set("dns_servers", dnses)
+
+	return nil
+}
+
+// resourceArmVirtualNetworkDelete deletes the specified ARM virtual network.
+func resourceArmVirtualNetworkDelete(d *schema.ResourceData, meta interface{}) error {
+	vnetClient := meta.(*ArmClient).vnetClient
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	_, err := vnetClient.Delete(resGroup, name)
+
+	return err
+}
+
+// getVirtualNetworkProperties is a helper function which returns the
+// VirtualNetworkPropertiesFormat of the network resource.
+func getVirtualNetworkProperties(d *schema.ResourceData) *network.VirtualNetworkPropertiesFormat {
+	// first; get address space prefixes:
+	prefixes := []string{}
+	for _, prefix := range d.Get("address_space").([]interface{}) {
+		prefixes = append(prefixes, prefix.(string))
+	}
+
+	// then; the dns servers:
+	dnses := []string{}
+	for _, dns := range d.Get("dns_servers").([]interface{}) {
+		dnses = append(dnses, dns.(string))
+	}
+
+	// then; the subnets:
+	subnets := []network.Subnet{}
+	if subs := d.Get("subnet").(*schema.Set); subs.Len() > 0 {
```

Comment:
Subnets on create or update appear to work OK (at least - the acceptance tests pass). I'm not sure it wouldn't actually be better to have separate resources for subnets though from a usability perspective.